[JSC][armv7] Fix YarrInterpreter alignment checks

eugeneia opened the pull request:
#### 6040dcdd450165720f3f141e3d07cbc3f0cd7f21
<pre>
[JSC][armv7] Fix YarrInterpreter alignment checks
<a href="https://bugs.webkit.org/show_bug.cgi?id=288079">https://bugs.webkit.org/show_bug.cgi?id=288079</a>

Reviewed by Yusuke Suzuki and Justin Michaud.

Compute required alignment in a platform independent way at compile time
using constexpr.

* Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
(JSC::Yarr::Interpreter::DisjunctionContext::allocationSize):
(JSC::Yarr::Interpreter::ParenthesesDisjunctionContext::allocationSize):

Canonical link: <a href="https://commits.webkit.org/291952@main">https://commits.webkit.org/291952@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/15b7d8d717626b626d190d124ed19db1b9580539

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93718 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13292 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3028 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98724 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44244 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13587 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21734 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71546 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28917 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96720 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10113 "Found 1 new test failure: fast/forms/ios/focus-input-via-button-no-scaling.html (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84702 "Found 1 new API test failure: TestWebKitAPI.WKWebExtensionAPIDevTools.MessagePassingFromPanelToBackground (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51880 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9795 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2329 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43559 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/86428 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80068 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2397 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100758 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/92384 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20770 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15154 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80560 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21022 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80644 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79898 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20066 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24448 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1800 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13926 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20754 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25932 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/115034 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20441 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33107 "Found 13 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.no-llint, microbenchmarks/memcpy-wasm-small.js.bytecode-cache, microbenchmarks/memcpy-wasm-small.js.default, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager-jettison, wasm.yaml/wasm/regress/llint-callee-saves-without-fast-memory.js.wasm-collect-continuously, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.default-wasm, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-eager-jettison, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-collect-continuously, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-collect-continuously ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23901 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22182 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -115,9 +115,9 @@ class Interpreter {
         static size_t allocationSize(unsigned numberOfFrames)
         {
-            static_assert(alignof(DisjunctionContext) <= sizeof(void*));
+            static_assert(alignof(DisjunctionContext) <= alignof(max_align_t));
```
Maybe we should leave a comment here about the reasoning. I.e, this is the guaranteed alignment from malloc

justinmichaud reviewed (approved):
Let's see what @msaboff thinks before landing

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -115,9 +115,9 @@ class Interpreter {
+            static_assert(alignof(DisjunctionContext) <= alignof(max_align_t));
             size_t rawSize = sizeof(DisjunctionContext) - sizeof(uintptr_t) + Checked<size_t>(numberOfFrames) * sizeof(uintptr_t);
-            size_t roundedSize = roundUpToMultipleOf<sizeof(void*)>(rawSize);
+            size_t roundedSize = roundUpToMultipleOf<alignof(max_align_t)>(rawSize);
```
It sounds wrong in 64bit environment as it becomes 16

justinmichaud replied:
My interpretation of this static assert was that we were only making sure that we didn't break the alignment guarantees of this class. Is there some other part of the static_assert that I am missing?

justinmichaud commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -110,14 +110,15 @@ class Interpreter {
 
         void* operator new(size_t, void* where)
         {
+            ASSERT(!((uintptr_t)where & (alignof(DisjunctionContext)-1)));
```
style nits

justinmichaud commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -110,14 +110,15 @@ class Interpreter {
         static size_t allocationSize(unsigned numberOfFrames)
         {
-            static_assert(alignof(DisjunctionContext) <= sizeof(void*));
+            static_assert(alignof(DisjunctionContext) <= sizeof(int64_t));
```
This is not really capturing the intent. I think this assert should be alignof(void*) || is32bit(), and the assert above should be 32-bit only. That is, on 64-bit we don't need the assert because malloc will give us memory with the proper alignment. We should  also leave a comment.

eugeneia replied:
I think the intent is as follows(?):

- `BumpPointerPool::alloc` does not align at all (an empty pool starts as page aligned though? `PageAllocation::allocate(poolSize)`)
- but we need DisjunctionContext to be aligned properly
- so what we do is we pad at the end of each allocation to ensure alignment

`size_t roundedSize = roundUpToMultipleOf<sizeof(void*)>(rawSize);`

- what we pad to here is arbitrarily `sizeof(void*)`, chosen to be >= our needed alignment while being as little overhead as possible
- the static assert checks that our chosen padding satisfies the correctness invariant ">= our needed alignment"

That's why I think the correct fix is explicitly padding to and checking `sizeof(int64_t)` (or really maybe name it `auto alignment = 8`). I don't think `alignof(void*)` makes sense because it asks the C compiler about what it expects, while really we manually control the (minimum) alignment in `allocDisjunctionContext`.

Begs the question why we don't just

`size_t roundedSize = roundUpToMultipleOf<alignof(DisjunctionContext)>(rawSize);`

and do without the static_assert since `alignof(DisjunctionContext) <= ourChosenConstant`?

eugeneia replied:
Hum maybe we want to make sure that `alignof(DisjunctionContext)` is small, hence we choose a small limit and do the static assert as a canary in case it ever somehow grows from a later change? Dunno I'm not super familiar with what kind of changes would bloat C++ alignof requirements.

eugeneia replied:
Actually, I believe sizeof(void*) is chosen so that the trailing `uintptr_t` frames are pointer aligned.

I have pushed a version that I think is correct on both 32 and 54 bit platforms and communicates the intent as I understand it.

msaboff replied:
Sorry for the long delay.  @eugeneia The change you did in [91952@main](https://commits.webkit.org/291952@main) looks good.  Yes, the intent of the original change was as you said with your comment "Actually, I believe sizeof(void*) is chosen so that the trailing uintptr_t frames are pointer aligned."

eugeneia committed:
[JSC][armv7] Fix YarrInterpreter alignment checks
https://bugs.webkit.org/show_bug.cgi?id=288079

Reviewed by Yusuke Suzuki and Justin Michaud.

Compute required alignment in a platform independent way at compile time
using constexpr.

* Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
(JSC::Yarr::Interpreter::DisjunctionContext::allocationSize):
(JSC::Yarr::Interpreter::ParenthesesDisjunctionContext::allocationSize):

Canonical link: https://commits.webkit.org/291952@main
